### PR TITLE
feat(core): update connector_mandate_details in payment_method

### PR DESCRIPTION
**Pull request by Aprabhat19**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

The connector_mandate_details, were stored in the db earlier. In this PR if two mandates are created with two different connectors, but for same payment_method, We would update the PaymentMethod entry in the db. Also if the expiry was changed then also we would  update connector_mandate_details.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
 
 - Create a MA and an MCA_1 with cyber source.
 
 `Scenario 1`
 - Make an off session payment
 -> `Create`
 ```
 curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_BgXaGEnYjJ4qj7ko3kJ7Rj40BnLYhYT4IEryQ5f2HPnGJhrW59GkdRUKqHWC2jdy' \
--data-raw '{
    "amount": 1000,
    "currency": "USD",
    "confirm": false,
    
    "customer_id": "new-c7",
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "authentication_type": "no_three_ds",
    "return_url": "https://google.com",
     "email": "something@gmail.com",
    "name": "Joseph Doe",
    "phone": "999999999",
    "phone_country_code": "+65",
    "description": "Its my first payment request",
"setup_future_usage":"off_session",
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },

    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "connector_metadata": {
        "noon": {
            "order_category": "pay"
        }
    }
}'
```
-> `Confirm`
```
curl --location 'http://localhost:8080/payments/pay_eWzzcEBGyPDdSqmqr7qE/confirm' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_BgXaGEnYjJ4qj7ko3kJ7Rj40BnLYhYT4IEryQ5f2HPnGJhrW59GkdRUKqHWC2jdy' \
--data '{
    "confirm": true,
       "payment_method": "card",
    "payment_method_type": "credit",
    
    "payment_method_data": {
        "card": {
             "card_number": "4111111111111111",
            
            
            "card_exp_month": "12",
            "card_exp_year": "2043",
            "card_holder_name": "Cy1",
            "card_cvc": "123"
        }
    },
  
      "customer_acceptance": {
            "acceptance_type": "offline",
            "accepted_at": "1963-05-03T04:07:52.723Z",
            "online": {
                "ip_address": "in sit",
                "user_agent": "amet irure esse"
            }}
}'
```
- A entry would be made in the connector_mandate_details with its respective mca_id
Screenshot 2024-03-28 at 12 55 01 PM

`Scenario 2`

- Create an MCA_2 with cyber source for the same profile, just change the `"connector_label":"cyber source_default_2"`
-Make another off_session payment, with the same payment_method
 -> `Create`
 ```
 curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_BgXaGEnYjJ4qj7ko3kJ7Rj40BnLYhYT4IEryQ5f2HPnGJhrW59GkdRUKqHWC2jdy' \
--data-raw '{
    "amount": 1000,
    "currency": "USD",
    "confirm": false,
    
    "customer_id": "new-c7",
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "authentication_type": "no_three_ds",
    "return_url": "https://google.com",
     "email": "something@gmail.com",
    "name": "Joseph Doe",
    "phone": "999999999",
    "phone_country_code": "+65",
    "description": "Its my first payment request",
"setup_future_usage":"off_session",
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },

    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "connector_metadata": {
        "noon": {
            "order_category": "pay"
        }
    }
}'
```

-> Confirm
```
curl --location 'http://localhost:8080/payments/pay_eWzzcEBGyPDdSqmqr7qE/confirm' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_BgXaGEnYjJ4qj7ko3kJ7Rj40BnLYhYT4IEryQ5f2HPnGJhrW59GkdRUKqHWC2jdy' \
--data '{
    "confirm": true,
   
          "payment_method": "card",
    "payment_method_type": "credit",
    
    "payment_method_data": {
        "card": {
             "card_number": "4111111111111111",
            
            
            "card_exp_month": "12",
            "card_exp_year": "2043",
            "card_holder_name": "Cy1",
            "card_cvc": "123"
        }
    },
        
            "routing": {
        "type": "single",
        "data": {"connector":"cybersource","merchant_connector_id":"mca_1GR9IUCJ08Ip7lbRMsG5"}
      },
      "customer_acceptance": {
            "acceptance_type": "offline",
            "accepted_at": "1963-05-03T04:07:52.723Z",
            "online": {
                "ip_address": "in sit",
                "user_agent": "amet irure esse"
            }}
}'
```
- A entry with updated connector_mandate_details is shown, with its respective `mca_id's` 
Screenshot 2024-03-28 at 12 54 52 PM

         `Scenario- 3`
- Make a payment with some Metadata-change, for eg:update the expiry number of the card`

 -> `Create`
 ```
 curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_BgXaGEnYjJ4qj7ko3kJ7Rj40BnLYhYT4IEryQ5f2HPnGJhrW59GkdRUKqHWC2jdy' \
--data-raw '{
    "amount": 1000,
    "currency": "USD",
    "confirm": false,
    
    "customer_id": "new-c7",
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "authentication_type": "no_three_ds",
    "return_url": "https://google.com",
     "email": "something@gmail.com",
    "name": "Joseph Doe",
    "phone": "999999999",
    "phone_country_code": "+65",
    "description": "Its my first payment request",
"setup_future_usage":"off_session",
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },

    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "connector_metadata": {
        "noon": {
            "order_category": "pay"
        }
    }
}'
```

-> `Confirm`
```
curl --location 'http://localhost:8080/payments/pay_eWzzcEBGyPDdSqmqr7qE/confirm' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_BgXaGEnYjJ4qj7ko3kJ7Rj40BnLYhYT4IEryQ5f2HPnGJhrW59GkdRUKqHWC2jdy' \
--data '{
    "confirm": true,
       "payment_method": "card",
    "payment_method_type": "credit",
    
    "payment_method_data": {
        "card": {
             "card_number": "4111111111111111",
            
            
            "card_exp_month": "10",
            "card_exp_year": "2043",
            "card_holder_name": "Cy1",
            "card_cvc": "123"
        }
    },
  
      "customer_acceptance": {
            "acceptance_type": "offline",
            "accepted_at": "1963-05-03T04:07:52.723Z",
            "online": {
                "ip_address": "in sit",
                "user_agent": "amet irure esse"
            }}
}'
```
- A updated entry of connector_mandate_id for that mca_id is shown
Screenshot 2024-03-28 at 12 57 01 PM

`Scenario- 4`
- Make a `off_session` payment with the same card .

 -> `Create`
 ```
 curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_BgXaGEnYjJ4qj7ko3kJ7Rj40BnLYhYT4IEryQ5f2HPnGJhrW59GkdRUKqHWC2jdy' \
--data-raw '{
    "amount": 1000,
    "currency": "USD",
    "confirm": false,
    
    "customer_id": "new-c7",
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "authentication_type": "no_three_ds",
    "return_url": "https://google.com",
     "email": "something@gmail.com",
    "name": "Joseph Doe",
    "phone": "999999999",
    "phone_country_code": "+65",
    "description": "Its my first payment request",
"setup_future_usage":"off_session",
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },

    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "connector_metadata": {
        "noon": {
            "order_category": "pay"
        }
    }
}'
```

-> `Confirm`
```
curl --location 'http://localhost:8080/payments/pay_eWzzcEBGyPDdSqmqr7qE/confirm' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_BgXaGEnYjJ4qj7ko3kJ7Rj40BnLYhYT4IEryQ5f2HPnGJhrW59GkdRUKqHWC2jdy' \
--data '{
    "confirm": true,
       "payment_method": "card",
    "payment_method_type": "credit",
    
    "payment_method_data": {
        "card": {
             "card_number": "4111111111111111",
            
            
            "card_exp_month": "10",
            "card_exp_year": "2043",
            "card_holder_name": "Cy1",
            "card_cvc": "123"
        }
    },
  
      "customer_acceptance": {
            "acceptance_type": "offline",
            "accepted_at": "1963-05-03T04:07:52.723Z",
            "online": {
                "ip_address": "in sit",
                "user_agent": "amet irure esse"
            }}
}'
```
- The entry would be updated with the new connector_mandate_details, with the same card

Screenshot 2024-03-28 at 12 58 04 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/build.rs [hunk 1]
```diff
@@ -2,7 +2,7 @@ fn main() {
     // Set thread stack size to 4 MiB for debug builds
     // Reference: https://doc.rust-lang.org/std/thread/#stack-size
     #[cfg(debug_assertions)]
-    println!("cargo:rustc-env=RUST_MIN_STACK=6291456"); // 6 * 1024 * 1024 = 6 MiB
+    println!("cargo:rustc-env=RUST_MIN_STACK= 6291456"); // 6 * 1024 * 1024 = 6 MiB
```
Comment (jagan-jaya) [inline crates/router/build.rs:5]: `space` needed here?

File: crates/router/src/core/payments/tokenization.rs [hunk 2]
```diff
@@ -89,19 +89,34 @@ where
...
+            };
+            let check_for_mit_mandates = resp.request.get_setup_mandate_details().is_none()
                 && resp
                     .request
                     .get_setup_future_usage()
                     .map(|future_usage| future_usage == storage_enums::FutureUsage::OffSession)
-                    .unwrap_or(false)
-            {
+                    .unwrap_or(false);
+            println!("ENTER");
```
Comment (jagan-jaya) [inline crates/router/src/core/payments/tokenization.rs:111]: remove this println

File: crates/router/src/core/payments/tokenization.rs [hunk 3]
```diff
@@ -112,6 +127,7 @@ where
             .attach_printable("Unable to serialize customer acceptance to value")?;
 
             let pm_id = if customer_acceptance.is_some() {
+                println!("ENTER HERE");
```
Comment (jagan-jaya) [inline crates/router/src/core/payments/tokenization.rs:130]: remove

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

File: crates/router/build.rs (hunk 1 above)
Comment (vspecky) [inline crates/router/build.rs:5]: Revert this change.

**Review by vspecky — CHANGES_REQUESTED** [review]
Please address mine and @jagan-jaya 's comments. Also please test this and add tests to the PR description.

**Review by jagan-jaya — APPROVED** [review]

**Review by Chethan-rao — APPROVED** [review]

**Pull request merged**